// aquaConfig/config.js
var fs = require('fs');

/*
 * Load the settings from the .json file on initialization.
 */
module.exports = function() {
    if (fs.existsSync('./templates/app/config.json')) { // Load the settings from the config file if it exists
        var cfg = require('../../app/config.json');

        if (cfg.pwmSettings) module.exports.pwmSettings = cfg.pwmSettings;
        if (cfg.controllerMode) module.exports.controllerMode = cfg.controllerMode;
        if (cfg.sunriseTime) module.exports.sunriseTime = cfg.sunriseTime;
        if (cfg.sunsetTime) module.exports.sunsetTime = cfg.sunsetTime;
        if (cfg.rampTime) module.exports.rampTime = cfg.rampTime;
        if (cfg.cloudFrequency) module.exports.cloudFrequency = cfg.cloudFrequency;
        if (cfg.lightningFrequency) module.exports.lightningFrequency = cfg.lightningFrequency;
        if (cfg.simulateMoon) module.exports.simulateMoon = cfg.simulateMoon;
        if (cfg.latitude) module.exports.latitude = cfg.latitude;
        if (cfg.longitude) module.exports.longitude = cfg.longitude;
        if (cfg.fanControl) module.exports.fanControl = cfg.fanControl;
        if (cfg.fanMax) module.exports.fanMax = cfg.fanMax;
    }

    return module.exports;
};

/*
 * Save the settings out to a .json file.
 */
module.exports.save = function() {
    fs.writeFileSync('./templates/app/config.json', JSON.stringify(module.exports));
    console.log('Saved config to "config.json"!')
};

/*
 * PWM Settings:
 *  Pulse-width modulation pin(s) for control of lights/fan.
 *
 *  Row 1 - Daytime max setting for each pin. (12 bit)
 *  Row 2 - Nighttime max setting for each pin. (12 bit)
 *  Row 3 - Delay/Offset settings, can use to delay certain pins when fading in or out. (Seconds)
 *  Row 4 - Channel, used for synchronizing pin(s) during light effects.
 *  Row 5 - Settings for channels during manual override events.
 *  Row 6 - BeagleBone Black pin(s) numbers.
 *
 *  12 bit range is 0 - 4095.
 */
module.exports.pwmSettings = [
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    ["P9_22", "P9_29", "P9_14", "P8_34", "P8_45", "P8_13", "P9_28", "P9_42"]
];

/*
 * Controller Mode
 *  Defines the mode that the controller runs in.
 *
 *  'auto' - Controller calculates the rise/set/phase times based on the
 *           configured location and date.
 *  'manual' - Controller uses the configured rise/set times and the max
 *             intensity for moonlight channels.
 */
module.exports.controllerMode = 'manual';

/*
 * Sunrise Time:
 *  Time when the lights attain full brightness in 'manual' mode.
 *  Stored in number of seconds from midnight.
 */
module.exports.sunriseTime = 25200;

/*
 * Sunset Time:
 *  Time when the lights begin to fade down in 'manual' mode.
 *  Stored in number of seconds from midnight.
 */
module.exports.sunsetTime = 64800;

/*
 * Ramp Time:
 *  Number of minutes of the fading period of lights.
 */
module.exports.rampTime = 45;

/*
 * Cloud Frequency:
 *  Fades lights to a lower brightness, to simulate a passing cloud.
 *
 *  0 - Feature disabled.
 *  1 - One cloud every 0 - 60 minutes. (random)
 *  2 - One cloud every 0 - 55 minutes. (random)
 *  ...
 *  13 - One cloud every minute.
 */
module.exports.cloudFrequency = 0;

/*
 * Lightning Frequency:
 *  Randomly generated lightning effects.  Occurs only when a cloud is overhead.
 *
 *  0 - Feature disabled.
 *  1 - 10% of clouds have a lightning effect.
 *  ...
 *  10 - 100% of clouds have a lightning effect.
 */
module.exports.lightningFrequency = 0;

/*
 * Simulate Moon:
 *  Enables adjusting of nighttime brightness to correspond to the phase and
 *  rise/set times of the Moon.  Only available in 'auto' mode.
 *
 *  0 - Feature disabled.
 *  1 - Feature enabled.
 */
module.exports.simulateMoon = 0;

/*
 * Latitude:
 *  Geographical latitude of the location to simulate rise/set/phase times for.
 *  Only used in 'auto' mode.
 */
module.exports.latitude = 0;

/*
 * Longitude:
 *  Geographical longitude of the location to simulate rise/set/phase times for.
 *  Only used in 'auto' mode.
 */
module.exports.longitude = 0;

/*
 * Fan Control:
 *  Designate pin number P9_42 to control a PWM PC Fan.
 *
 *  0 - Feature disabled.
 *  1 - Feature enabled.
 */
module.exports.fanControl = 0;

/*
 * Fan Max:
 *  The maximum setting for the fan speed.
 *  Only used if 'fanControl' is enabled.
 */
module.exports.fanMax = 0;

/**
 * Specifies pin number to watch for manual override events.
 * Will turn the light to the manual override settings when this pin is switched HIGH.
 * @type {string}:BeagleBone Black pin number.
 */
module.exports.manualOverridePin = "";
